src/bin/sse-test-api/main.rs

```diff
@@ -0,0 +1,195 @@
+    stream_url: String,
+    /// The URL of a callback endpoint created by the test harness .
+    callback_url: String,
+    /// An optional integer specifying the initial reconnection delay parameter, in
+    /// milliseconds. Not all SSE client implementations allow this to be configured, but the
+    /// test harness will send a value anyway in an attempt to avoid having reconnection tests
+    /// run unnecessarily slowly.
+    initial_delay_ms: Option<u64>,
+    /// An optional integer specifying the read timeout for the connection, in
+    /// milliseconds.
+    read_timeout_ms: Option<u64>,
+    /// A JSON object containing additional HTTP header names and string values. The SSE
+    /// client should be configured to add these headers to its HTTP requests; the test harness
+    /// will then verify that it receives those headers. The test harness will only set this
+    /// property if the test service has the "headers" capability. Header names can be assumed
+    /// to all be lowercase.
+    headers: Option<HashMap<String, String>>,
+}
+
+#[derive(Serialize, Debug)]
+#[serde(tag = "kind")]
+enum EventType {
+    #[serde(rename = "event")]
+    Event { event: Event },
+    #[serde(rename = "error")]
+    Error { error: String },
+}
+
+impl From<es::Event> for EventType {
+    fn from(event: es::Event) -> Self {
+        Self::Event {
+            event: Event {
+                event_type: event.event_type.clone(),
+                data: String::from_utf8(event.field("data").unwrap_or_default().to_vec()).unwrap(),
+                id: String::from_utf8(event.field("id").unwrap_or_default().to_vec()).unwrap(),
+            },
+        }
+    }
+}
+
+#[derive(Serialize, Debug)]
+struct Event {
+    #[serde(rename = "type")]
+    event_type: String,
+    data: String,
+    id: String,
+}
+
+async fn status() -> impl Responder {
+    web::Json(Status {
+        capabilities: vec![
+            "comments".to_string(),
+            "retry".to_string(),
+            "post".to_string(),
+            "report".to_string(),
+            "read-timeout".to_string(),
+            "headers".to_string(),
+            "last-event-id".to_string(),
+        ],
+    })
+}
+
+async fn stream(
+    req: HttpRequest,
+    config: web::Json<Config>,
+    app_state: web::Data<AppState>,
+) -> HttpResponse {
+    let mut stream_entity = match StreamEntity::new(config.into_inner()) {
+        Ok(se) => se,
+        Err(e) => return HttpResponse::InternalServerError().body(e),
+    };
+
+    let mut counter = match app_state.counter.lock() {
+        Ok(c) => c,
+        Err(_) => return HttpResponse::InternalServerError().body("Unable to retrieve counter"),
+    };
+
+    let mut entities = match app_state.stream_entities.lock() {
+        Ok(h) => h,
+        Err(_) => return HttpResponse::InternalServerError().body("Unable to retrieve handles"),
+    };
+
+    *counter += 1;
+    stream_entity.start();
+    entities.insert(*counter, stream_entity);
```

Comment:
We should move these down below the match since generating the URL could theoretically fail.